docs/source/dataset_streaming.rst

```diff
@@ -0,0 +1,138 @@
+Load a Dataset in Streaming mode
+==============================================================
+
+When a dataset is in streaming mode, you can iterate over it directly, without having to download the entire dataset.
+The data are downloaded progressively as you iterate over the dataset.
+You can enable dataset streaming by passing ``streaming=True`` in the :func:`load_dataset` function to get an iterable dataset.
+
+This is useful if you don't have enough space on your disk to download the dataset, or if you don't want to wait for your dataset to be downloaded before using it.
+
+Here is a demonstration:
+
+.. code-block::
+
+    >>> from datasets import load_dataset
+    >>> dataset = load_dataset('oscar', "unshuffled_deduplicated_en", split='train', streaming=True)
+    >>> print(next(iter(dataset)))
+    {'text': 'Mtendere Village was inspired by the vision of Chief Napoleon Dzombe, which he shared with John Blanchard during his first visit to Malawi. Chief Napoleon conveyed the desperate need for a program to intervene and care for the orphans and vulnerable children (OVC) in Malawi, and John committed to help...
+
+Even though the dataset is 1.2 terabytes of data, you can start using it right away. Under the hood, it downloaded only the first examples of the dataset
+
+.. note::
+
+    The dataset that is returned is a :class:`datasets.IterableDataset`, not the classic map-style :class:`datasets.Dataset`. To get examples from an iterable dataset, you have to iterate over it using a for loop for example. To get the very last example of the dataset, you first have to iterate on all the previous examples.
```

Comment:
is it possible to create a classic `Dataset` from an `IterableDataset`? 

one application that i have in mind is picking the first N examples of a huge dataset, collecting them in a standard `Dataset` and then doing all my exploration / preprocessing / task preparation etc on that dataset.

e.g. something like

```python
from datasets import load_dataset 

dataset = load_dataset('oscar', "unshuffled_deduplicated_en", split='train', streaming=True)
# create a `Dataset` i can play with?
sample = dataset.select(range(100))
```

Reply:
Sure definitely :)

I was thinking of adding something like this in a next PR.
Maybe `IterableDataset.to_map_style_dataset()` ?
To get only the first examples we can also add `IterableDataset.take(n)`

Reply:
yeah, both those features would be great!

Reply:
I like the `IterableDataset.take(n)` as well. Could we also have a `IterableDataset.sample(n)` taking a random sample?

Reply:
a random sample would be very neat as well. here we might want to use something like reservoir sampling to deal with unbounded streams: https://en.wikipedia.org/wiki/Reservoir_sampling

Reply:
As soon as we have `.take()`, you can do `iterable_dataset.shuffle(buffer_size=buffer_size, seed=seed).take(n)` to take random samples.
This could be simplified by adding a `.sample()` method indeed

Reply:
How about slicing support i.e `iterable_dataset[100:200]` to get an iterator or `Dataset` at a particular slice?

Reply:
I would like to avoid allowing users to get items using `__getitem__` since it's not a map-style dataset.
So I agree it would be nice to get a slice of the data, but with a different API. Maybe something like
```python
sliced_dataset = iterable_dataset.skip(100).take(100)
```
What do you think ?

This is pretty close to the tf.data.Dataset API, which is also an iterable dataset.

Reply:
Yes, very cool @lhoestq